src/language-js/printer-estree.js

```diff
@@ -1808,7 +1808,21 @@ function printPathNoParens(path, options, print, args) {
         let res;
         if (isStringLiteral(n.value)) {
           const value = rawText(n.value);
-          res = '"' + value.slice(1, -1).replace(/"/g, "&quot;") + '"';
+          const innerValue = value.slice(1, -1);
+          const hasSingle = innerValue.includes("'");
+          const hasDouble = innerValue.includes('"');
+
+          if (options.singleQuote === "all") {
+            if (hasSingle && hasDouble) {
```

Comment:
I think we should count the number of singles and doubles and choose surrounding quotes so that the number of escapes is minimized, like we do for JS strings.

But wait a minute – can there ever be both `'`s and `"`s at the same time in an attribute value? I don't think so. And is there are risk of getting all `'`s and `"`s turned into `&apos;` and `&quot;` as you run Prettier in between edits? Should we unescape unnecessary `&apos;` and `&quot;`?

Reply:
valid point, I just realized as well that it's not valid to escape quotes in pure JSX attributes, i.e.

this is invalid: `<div id="invalid \" jsx" />`
this works though: `<div id={"valid \" jsx"} />`

I'm updating the format to use the same logic from `printString`, [there's a previous thread going through this discussion in more depth above](https://github.com/prettier/prettier/pull/4798#discussion_r199756260)

Reply:
It's not invalid; it just won't do what you mean 😄

It's better to just convert to string expressions if you need any escaping.